### ISSUE TITLE
$this->Auth->user()

### ISSUE THREAD

**Issue by MatheusVellone** [ADmad/CakePHP-HybridAuth#15]
After a successfull login, the info about the user should be on $this->Auth->user() or not ?
I guess not because mine isn't. 

I did some research on the plugin and library code, and found the info on $_SESSION["HA::STORE"]['hauth_session.provider.user']. You plan put it on Auth->user() ? I tried to do this, but i don't understand very well about Components.

Sry for bad english

**Comment by ADmad** [comment]
This plugins provides an authenticate class which like any other core authentication class provides a user record which is available through `$this->Auth->user()` after successful authentication.

**Comment by MatheusVellone** [comment]
Well, mine isn't working sry.
Can you please tell me where exactly in the code this happens ? Where you set the info that can be retrieved by $this->Auth->user() ?

**Comment by MatheusVellone** [comment]
Ok, it works now. But the user must click on the "login with..." twice to work. That might be some implementation error. When I found it, i'll post here in case someone else have the same error

**Comment by ADmad** [comment]
> But the user must click on the "login with..." twice to work.

This sounds similar to #8. It's a redirection issue, the login is done first time but you are redirected back to login page instead of `loginRedirect` URL. But I have been unable to reproduce this problem.

**Comment by fazalmohammad19** [comment]
Somehow you are missing to understand the problem. You said  "you are redirected back to login page" ---Is this line said by above user (MatheusVellone )? Actually it's not sending to login page. Its sending to redirectUrl (that might be anything -login or home etc) Its redirecting properly (here $this->Auth->identify() is null ). But after redirecting you have to click on login url. (it might be in a buton or as a link , anywhere in the redirected page). This 2nd time click makes the login accurately. (this means we get $this->Auth->identify() value)

You are thinking the issue in a shape of your mind. As in   #8  said and also i am repeating here- Please create a blank project you will see the issue.  Use twitter as provider

**Comment by ADmad** [comment]
> Its sending to redirectUrl (that might be anything -login or home etc) Its redirecting properly (here $this->Auth->identify() is null ).

The `redirectUrl` is normally set to a page which is accessible only after login. The url returned by `redirectUrl()` method can either be referrer or value of `AuthComponent::$redirectUrl` or '/'. Never login  url itself. If as you say you are properly redirect back but not logged in, it means that url is publicly accessible. Auth won't check for authentication if you access a publicly accessible url. At this point instead of going to login page just to go any protected page and you will be logged in and able to access the page. Confirm that this is the behavior you get and I will explain why.

> As in #8 said and also i am repeating here- Please create a blank project in cake 3, you will see the issue.

I have already checked using a fresh cake 2 app (cake version wasnt specified in any of the tickets) wit 2 providers  and don't have any problem. I will test with cake 3 too.

You can also catch me on IRC if you want to discuss the issue.

**Comment by MatheusVellone** [comment]
If i put the `loginRedirect` a page that's only acessible if there's a user logged in, after the first loggi attempt it redirect to this page and i can access it normally. So I'm logged in, but the needed info isn't accessible by `$this->Auth->user()`. I'm running a few tests here based on the comments

**Comment by ADmad** [comment]
> If i put the loginRedirect a page that's only acessible if there's a user logged in, after the first loggi attempt it redirect to this page and i can access it normally. So I'm logged in, but the needed info isn't accessible by $this->Auth->user().

I am baffled by this. If you can access a protected page it means you are logged in and Auth->user() should return the info.

**Comment by MatheusVellone** [comment]
So do I. This in #8 makes sense to me.

> (to actually log you in the first time) because when it comes back from facebook through the endpoint it's not a post anymore and that's why when you press the button to login with facebook a second time, it will work because it already has the facebook stuff stored in session and then its actually a post.

But, i can assure you the info isn't there for the first time.

**Comment by ADmad** [comment]
Once you login with the provider you are redirected back to HybridAuthController::endpoint() which calls HybridAuth::endpoint() and the user credentials get stored in hyridauth's session store. 

After that hyridauth redirects to the url it has in `hauth_return_to`. Here's where the issue arises. At this point hyridauth has the user credentials but Auth does not. Whether the info is synced with Auth or not depends on what url you are redirected to. If that page is a protected page Auth attempts to get user info and HyridAuthenticate::getUser() fetch that info from hyridauth and returns it to Auth component and everything is hunky dory. 

But if the page you are redirected is publicly accessible page Auth won't attempt to get user info which is why you guys think the login through provider has not succeeded. As soon as you access a protected page Auth will get the user info from hybridauth.

So the solution to the issue would be add some kind of check in AppController::beforeFilter() if hybridauth already has the user info and force Auth to get that info. Or to keep things simpler for devs i will have to add another action in HyridAuthController and always redirected to that action after login through provider to ensure the user info from hyridauth is synced with Auth component and then redirect to whatever url `AuthComponent::redirectUrl()` returned.

**Comment by MatheusVellone** [comment]
> add another action in HyridAuthController and always redirected to that action after login through provider to ensure the user info from hyridauth is synced with Auth component and then redirect to whatever url AuthComponent::redirectUrl() returned.

I would be grateful! If you want me to test any code before you merge into the master branch, just tell me

**Comment by ADmad** [comment]
@MatheusVellone Are you using cake2 or cake3 ?

**Comment by fazalmohammad19** [comment]
@ADmad  , Now you have reached to the point. Your library was always working well. It was always authenticate properly but for the 1st initiation it was not sync with cakephp auth user (as i mention previouslly $this->Auth->identify() is null ).  

adding another action in HyridAuthController is best. or this issue will be raised by every new user

Thanks

**Comment by MatheusVellone** [comment]
@ADmad Cake 2

**Comment by ADmad** [comment]
Adding another action in HyridAuthController would add an extra round trip to the app :(   Not to happy about that but will do that for now until i come up with a better solution. Don't want to deal with this issue again :)

**Comment by ADmad** [comment]
Adding an extra action might not be the best way to deal with this. If I do that devs will also have to ensure their authorization scheme allows all users to access that action.

**Comment by MatheusVellone** [comment]
Here's something weird. I found this on hybridauth.log
`DEBUG -- ::1 -- 2015-03-06T18:30:16+01:00 -- Hybrid_Auth::setup( Steam ). HybridAuth Callback URL set to:  -- http://localhost/mysite/users/logout`
This log message is set on the hybridAuth plugin (not the @ADmad's one)
>`$params["hauth_return_to"] = Hybrid_Auth::getCurrentUrl();`
>`Hybrid_Logger::debug("Hybrid_Auth::setup( $providerId ). HybridAuth Callback URL set to: ", $params["hauth_return_to"]);`
idk how it gets users/logout as currentUrl

**Comment by MatheusVellone** [comment]
I put a `die(debug($adapter));` [here](https://github.com/ADmad/CakePHP-HybridAuth/blob/master/Controller/Component/Auth/HybridAuthAuthenticate.php#L111)
On the first time it doesn't enter there, but does on the second (when the info is set).
When the page is redirected to the provider's login page, doesn't this code stops ? On the first attempt.
But on the second, when there's info on the session, it skips to aquire the adapter and does the trick.

**Comment by ADmad** [comment]
@MatheusVellone Try using `hauth-redirect` branch. (`dev-hauth-redirect` for composer).

**Comment by MatheusVellone** [comment]
Oh dude! You're a genius! haha
It worked on first attempt, just the redirect is wrong. It took me to http://localhost/scf/hybrid_auth/Home/couponList instead of http://localhost/scf/Home/couponList

**Comment by ADmad** [comment]
Set 'plugin' => false in your loginRedirect url.

**Comment by MatheusVellone** [comment]
Yeah. Now it works very well! :+1:
Thanks for the help, for the plugin and sorry for bothering you
